### PR TITLE
Vin/mags improvements

### PR DESCRIPTION
**Pull request by vinnie4k**
## Overview

This PR provides improvements and modifications for magazines.

### Magazines List

- Refactored `MagazinesList` to support MVVM
    - Created a `ViewModel` class and separated logic from UI
- Users can now view magazines from “All semesters” (temporary limit of 15)
    - Wrote a GraphQL query for fetching all magazines
- Reversed the order displayed in the semester dropdown with the most recent at the top

### Weekly Debrief

- Removed the “top 10% active readers” on weekly debrief as requested

### Magazine Reader

- Added a `ProgressView` that will spin if a magazine is loading
- Fixed bug where sharing a magazine caused the app to crash

### Minor Changes

- Updated `Secrets.plist` for magazines

## Test Coverage

- Used different devices with varying screen sizes in Simulator to see if there were any UI issues
- In-depth playtesting to minimize any bugs that may be present

## Next Steps

- Fetching all magazines has a limit of 15 at this moment. Once we get more magazines, will need to implement pagination/infinite scrolling for magazines.
- Add magazines to weekly debrief.
- Creating a cache system for magazines may help with performance and prevent repeated downloads of PDFs.

## Screenshots

<details>
  <summary>Weekly Debrief</summary>
  
</details>
<details>
  <summary>Magazines List</summary>
  
</details>

### CODE REVIEW

File: Volume/Supporting Views/SemesterMenuView.swift [hunk 1]
```diff
@@ -25,7 +25,7 @@ struct SemesterMenuView: View {
                 Text(Self.format(semesterString: selection ?? options.last ?? Constants.semesterPlaceholder))
                     .font(.helveticaNeueMedium(size: Constants.fontSize))
                     .fixedSize()
-
+
```
Comment (ReadeP) [inline Volume/Supporting Views/SemesterMenuView.swift:28]: remove tabbing

File: Volume/Supporting Views/SemesterMenuView.swift [hunk 2]
```diff
@@ -46,6 +46,8 @@ struct SemesterMenuView: View {
 extension SemesterMenuView {
 
     private static func format(semesterString: String) -> String {
+        if (semesterString == "all") { return "All semesters" }
```
Comment (ReadeP) [inline Volume/Supporting Views/SemesterMenuView.swift:49]: remove parentheses

File: Volume/View Models/MagazinesListViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        @Published var queryBag = Set<AnyCancellable>()
+                
+        var networkState: NetworkState = NetworkState()
+        private var numMagsLoaded: Double = 0
+        private var offset: Double = 0
+
+        private struct Constants {
+            static let allMagazinesLimit: Double = 15
+            static let featuredMagazinesLimit: Double = 7
+            static let animationDuration = 0.1
```
Comment (ReadeP) [inline Volume/View Models/MagazinesListViewModel.swift:31]: Let's add `: Double` to make this look parallel

**Review by ReadeP — APPROVED** [review]
Great work Vin! Looks good style wise. Please wait for Vivian, Jennifer, or Han to give you more feedback on the SwiftUI components

File: Volume/View Models/MagazinesListViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+            )
+            .map { $0.magazines.map(\.semester) }
+            .sink { [weak self] completion in
+                self?.networkState.handleCompletion(screen: .magazines, completion)
+            } receiveValue: { [weak self] semesters in
+                guard let `self` = self else { return }
+                self.allSemesters = Array(Set(semesters))
+                    .sorted(by: self.compareSemesters)
+                
+                self.allSemesters?.reverse()
```
Comment (jjennifergu) [inline Volume/View Models/MagazinesListViewModel.swift:74]: can just change the definition of `compareSemesters` to be reversed

File: Volume/View Models/MagazinesListViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        // MARK: Helpers
+
+        private func compareSemesters(_ s1: String, _ s2: String) -> Bool {
+            if let s1Year = Int(s1.suffix(2)),
+               let s2Year = Int(s2.suffix(2)),
+               s1Year != s2Year {
+                return s1Year < s2Year
+            }
+
+            let validSemesterStrings = ["sp", "fa"]
```
Comment (jjennifergu) [inline Volume/View Models/MagazinesListViewModel.swift:128]: consider making this into an enum too, and change the switch statement below accordingly

**Review by jjennifergu — APPROVED** [review]
Great job Vin! Just suggested some small changes but overall this is really solid :)

File: Volume/View Models/MagazinesListViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+            sectionQueries.magazinesBySemester = Network.shared.publisher(
+                for: GetMagazinesBySemesterQuery(semester: semester)
+            )
+            .map { $0.magazines.map(\.fragments.magazineFields) }
+            .sink { [weak self] completion in
+                self?.networkState.handleCompletion(screen: .magazines, completion)
+            } receiveValue: { [weak self] magazineFields in
+                Task {
+                    let magazinesBySemester = await [Magazine](magazineFields)
+                    withAnimation(.linear(duration: 0.1)) {
```
Comment (vianxnguyen) [inline Volume/View Models/MagazinesListViewModel.swift:94]: can use Constants.animationDuration here

File: Volume/View Models/MagazinesListViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        featuredMagazines: MainView.TabState<[Magazine]>,
+        magazinesBySemester: MainView.TabState<[Magazine]>
+    )
+    
+    typealias SectionQueries = (
+        featuredMagazines: AnyCancellable?,
+        magazinesBySemester: AnyCancellable?
+    )
+
+    private static func format(semesterString: String) -> String {
```
Comment (vianxnguyen) [inline Volume/View Models/MagazinesListViewModel.swift:157]: i'd factor out these string/comparison helpers in separate supporting file

File: Volume/View Models/MagazinesListViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        @Published var deeplinkId: String?
+        @Published var openMagazineFromDeeplink = false
+        @Published var sectionQueries: SectionQueries = (nil, nil)
+        @Published var sectionStates: SectionStates = (.loading, .loading)
+        @Published var selectedSemester: String?
+        @Published var allSemesters: [String]? = nil
+        @Published var queryBag = Set<AnyCancellable>()
+                
+        var networkState: NetworkState = NetworkState()
+        private var numMagsLoaded: Double = 0
```
Comment (vianxnguyen) [inline Volume/View Models/MagazinesListViewModel.swift:25]: how is numMagsLoaded used?

**Review by vianxnguyen — APPROVED** [review]
looks amazing!! 👍

File: Volume/Views/MainView/MagazinesList.swift [hunk 9]
```diff
@@ -2,105 +2,32 @@
 //  MagazinesList.swift
 //  Volume
 //
-//  Created by Hanzheng Li on 3/4/22.
-//  Copyright © 2022 Cornell AppDev. All rights reserved.
+//  Created by Vin Bui on 3/18/23.
+//  Copyright © 2023 Cornell AppDev. All rights reserved.
```
Comment (dishanest) [inline Volume/Views/MainView/MagazinesList.swift:6]: the disrespect 😂

File: Volume/Views/MainView/MagazinesList.swift [hunk 10]
```diff
@@ -2,105 +2,32 @@
...
 //
 
-import Combine
 import SwiftUI
+import Combine
```
Comment (dishanest) [inline Volume/Views/MainView/MagazinesList.swift:10]: alphabetize imports

**Review by dishanest — COMMENTED** [review]
Good work! Just a quick reminder to use as much previously implemented logic as possible so you don't have to duplicate logic/code. You'll get a better hang of it as you become more familiar w/ the repo. 

For the PR description, make sure to include both before and after screenshots!

File: Volume/Views/MainView/MagazinesList.swift (hunk 9 above)
Comment (vinnie4k) [reply Volume/Views/MainView/MagazinesList.swift:6]: i rewrote a lot of stuff so whoever works on this next can blame me if it doens't work

**Pull request merged**